### PR TITLE
Fall back to untranslated string rather than showing missing translation error

### PR DESCRIPTION
**Pull request by novocaine**
When a translation string is not found, even in the fallback dictionary (en), rather than showing a translation error message, show the untranslated string. This is less bad than showing the error as it provides an english fallback, albeit a potentially cryptic one if the input involves un-interpolated placeholders.

This situation comes about when translation resource files fail to load from the server, but also during dev if the developer fails to generate an entry in the translations dictionary by running `yarn i18n`. So disable the behaviour in development. 

Suggested in https://github.com/vector-im/element-web/issues/9422#issuecomment-1083294513 and this is also the behaviour of GNU gettext.

---
Here's what your changelog entry will look like:

## ✨ Features
 * Fall back to untranslated string rather than showing missing translation error ([\#8609](https://github.com/matrix-org/matrix-react-sdk/pull/8609)). Contributed by @novocaine.

### CODE REVIEW

File: src/languageHandler.tsx [hunk 1]
```diff
@@ -90,6 +90,22 @@ const translateWithFallback = (text: string, options?: object): { translated?: s
     const translated = counterpart.translate(text, { ...options, fallbackLocale: counterpart.getLocale() });
     if (!translated || /^missing translation:/.test(translated)) {
         const fallbackTranslated = counterpart.translate(text, { ...options, locale: FALLBACK_LOCALE });
+        if (!fallbackTranslated || /^missing translation:/.test(translated) && process.env.NODE_ENV !== "development") {
```
Comment (t3chguy) [inline src/languageHandler.tsx:93]: Why a regex over `translated.startsWith(...)`?

**Review by t3chguy — APPROVED** [review]
LGTM otherwise

File: src/languageHandler.tsx (hunk 1 above)
Comment (novocaine) [reply src/languageHandler.tsx:93]: consistency with the existing code (I guess that might predate startsWith)

Comment (novocaine) [reply src/languageHandler.tsx]: I've replaced with `startsWith` which I think is fine, and along the way realised the second test was erroneously against `translated`, not `fallbackTranslated` ..

File: src/languageHandler.tsx [hunk 2]
```diff
@@ -88,8 +88,25 @@ export function _td(s: string): string { // eslint-disable-line @typescript-esli
  * */
 const translateWithFallback = (text: string, options?: object): { translated?: string, isFallback?: boolean } => {
     const translated = counterpart.translate(text, { ...options, fallbackLocale: counterpart.getLocale() });
-    if (!translated || /^missing translation:/.test(translated)) {
+    if (!translated || translated.startsWith("missing translation:")) {
         const fallbackTranslated = counterpart.translate(text, { ...options, locale: FALLBACK_LOCALE });
+        if (!fallbackTranslated || fallbackTranslated.startsWith("missing translation:")
+            && process.env.NODE_ENV !== "development") {
```
Comment (t3chguy) [inline src/languageHandler.tsx:94]:
```suggestion
        if (!fallbackTranslated ||
            (fallbackTranslated.startsWith("missing translation:") && process.env.NODE_ENV !== "development")
        ) {
```

To save someone needing to check up the operator precedence

File: src/languageHandler.tsx [hunk 3]
```diff
@@ -88,8 +88,25 @@ export function _td(s: string): string { // eslint-disable-line @typescript-esli
  * */
 const translateWithFallback = (text: string, options?: object): { translated?: string, isFallback?: boolean } => {
     const translated = counterpart.translate(text, { ...options, fallbackLocale: counterpart.getLocale() });
-    if (!translated || /^missing translation:/.test(translated)) {
+    if (!translated || translated.startsWith("missing translation:")) {
         const fallbackTranslated = counterpart.translate(text, { ...options, locale: FALLBACK_LOCALE });
+        if ((!fallbackTranslated || fallbackTranslated.startsWith("missing translation:"))
```
Comment (t3chguy) [inline src/languageHandler.tsx:93]: Just to triple check intent, these brackets that were added actually changed the behaviour of the `if` - `&&` has higher precedence than `||`

Comment (novocaine) [reply src/languageHandler.tsx:93]: noted, the PR has been a litany of careless errors unfortunately. 

the intent is to never enter the if clause if we are in development mode (i.e. leave the behaviour as-is)

**Review by t3chguy — APPROVED** [review]

**Pull request merged**